lib/spack/spack/test/cmd/diff.py

```diff
@@ -0,0 +1,79 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import pytest
+
+import spack.cmd.diff
+import spack.config
+import spack.main
+import spack.store
+import spack.util.spack_json as sjson
+
+install = spack.main.SpackCommand('install')
+diff = spack.main.SpackCommand('diff')
+
+
+def test_diff(install_mockery, mock_fetch, mock_archive, mock_packages):
+    """Test that we can install two packages and diff them"""
+
+    specA = spack.spec.Spec('mpileaks').concretized()
+    specB = spack.spec.Spec('mpileaks+debug').concretized()
+
+    # Specs should be the same as themselves
+    c = spack.cmd.diff.compare_specs(specA, specA, to_string=True)
+    assert len(c['a_not_b']) == 0
+    assert len(c['b_not_a']) == 0
+
+    # Calculate the comparison (c)
+    c = spack.cmd.diff.compare_specs(specA, specB, to_string=True)
+    assert len(c['a_not_b']) == 1
+    assert len(c['b_not_a']) == 1
+    assert c['a_not_b'][0] == ['variant_value', 'mpileaks debug bool(False)']
+    assert c['b_not_a'][0] == ['variant_value', 'mpileaks debug bool(True)']
+
+
+def test_load_first(install_mockery, mock_fetch, mock_archive, mock_packages):
+    """Test with and without the --first option"""
+    install('mpileaks')
+
+    # Only one version of mpileaks will work
+    diff('mpileaks', 'mpileaks')
+
+    # 2 specs are required for a diff
+    with pytest.raises(spack.main.SpackCommandError):
+        diff('mpileaks')
+    with pytest.raises(spack.main.SpackCommandError):
+        diff('mpileaks', 'mpileaks', 'mpileaks')
+
+    # Ensure they are the same
+    assert "No differences" in diff('mpileaks', 'mpileaks')
+    output = diff('--json', 'mpileaks', 'mpileaks')
+    result = sjson.load(output)
+
+    assert len(result['a_not_b']) == 0
+    assert len(result['b_not_a']) == 0
+
+    assert 'mpileaks' in result['a_name']
+    assert 'mpileaks' in result['b_name']
+    assert "intersect" in result and len(result['intersect']) > 50
+
+    # After we install another version, it should ask us to disambiguate
+    install('mpileaks+debug')
+
+    # There are two versions of mpileaks
+    with pytest.raises(spack.main.SpackCommandError):
+        diff('mpileaks', 'mpileaks+debug')
+
+    # But if we tell it to use the first, it won't try to disambiguate
+    assert "variant" in diff('--first', 'mpileaks', 'mpileaks+debug')
+
+    # This matches them exactly
+    output = diff("--json", "mpileaks@2.3/ysubb76", "mpileaks@2.3/ft5qff3")
```

Comment:
@tgamblin @vsoch am I reading this correctly that these hashes are "magic numbers" hardcoded in? I think we need to change this to read the hash values from the installed specs. @nhanford is running into problems where this test is not stable across a change in hash algorithm or a change the mock packages.

Reply:
I don't know what you mean by magic numbers, but they refer to the same hashes you'd see on the command line with spack find and including "hash" for format:
```
$ bin/spack find --format "{name}/{hash}"
autoconf/5nymyrr2ar3userqwgewuayirjdo7c2w          ncurses/buo3tjmtuimvtofgncv6w3lzg2gmouqu
autoconf-archive/mplalc4sz2cys2wkwzji2ltyklv7x5xf  openssh/ddmvmwpu6feqyhtwnfxrsoflngvma7ll
automake/cy6bt5p74uc4qlj6jeyzuoiboaeipn5b          openssl/5m5cysfnhlgemdsngdbf4jb75cv73o7l
berkeley-db/m6cpl2epxybbnhypwvu7r3h7cq256uvg       openssl/gefi67ctqiegtpzvvnougob5oszdlysi
```
So I think if the core of how spack generates these hashes changes (@nhanford's work?) these would too and the test would also need to be updated.

Reply:
Yeah, I think as part of his work Nate is going to query spack for the hashes instead of having them hardcoded, so they don't have to be maintained going forward.

Reply:
I have a fix for the test over in #25310 (still on my branch but I will push it soon). I can probably push it to Nate’s branch.

Reply:
The issue in the test isn’t the hashes here but there are some assumptions about, e.g.,  the number of attributes that differ between specs. Those change in #25310 and I bet also in @nhanford’s PR.

Reply:
I fixed it in my PR and it should be pushed to my branch.

Reply:
I have been trying to rebase #21683 and encountered the same issue. I think this test is bound to fail whenever the hashes change? Would it be possible to recode the test in a way to be more resilient to that?

Reply:
We should have this fixed in #22845, but I need to fix a few things before we can merge that one.